text/0000-nested-publish.md

```diff
@@ -0,0 +1,303 @@
+
+```toml
+[package]
+name = "foo"
+version = "0.1.0"
+edition = "2021"
+publish = true
+
+[dependencies]
+foo-macros = { path = "macros", publish = "nested" } # new syntax
+```
+
+The sub-package manifest `foo/macros/Cargo.toml`:
+
+```toml
+[package]
+name = "macros"            # this name need not be claimed on crates.io
+# version = "0.0.0"        # version number is not used and may be omitted
+edition = "2021"
+publish = "nested"         # new syntax
+
+[lib]
+proc-macro = true
+```
+
+Then you can `cargo publish` from within the parent directory `foo/`, and this will create a single `foo` package on `crates.io`, with no `macros` (or `foo-macros`) package visible except when inspecting the source code or in compilation progress messages.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Cargo: Manifest
+
+Two new possible values are added to the manifest.
+
+* The `package.publish` field allows `"nested"` as a value, in addition to existing `false` and `true`. This value affects `cargo publish` and nested dependencies as described below.
+* The `dependencies.*.publish` field is newly defined, with the only currently allowed value being `"nested"`, to declare that that dependency is a nested dependency. (If desired, `publish = false` could be also be permitted to explicitly document an intent not to nest; this would be identical to the status quo and to omitting the key.)
+    * It is an error if a nested dependency does not have a `path` field, or if it has a `version`, `git`, or any other package source field, unless future work defines a meaning for that combination.
+
+When a nested dependency is present (making its referent be a nested package), the following additional requirements apply:
+
+* The nested package must have `package.publish = "nested"`; both `false` and `true` are errors.
+
+* If the nested package specifies `package.license`, the parent package's (not necessarily the dependent's) license expression must comply with the nested package's. This check is done solely in terms of the operators in the license expression. For example, if two nested packages contain licenses of `MIT` and `BSD-3-Clause`, then the parent package's expression must be `MIT AND BSD-3-Clause` or similar.
+
+  This check is intended only to prevent accidents (such as vendoring a third-party package without considering the implications of redistributing it). It is always valid to omit `package.license` from the nested package, thus making no machine-readable claims about its licensing independent from the parent package's.
+
+* We might want to explicitly prohibit nested packages from specifying a `package.version`, to avoid giving the misleading impression that it means anything. (`package.version` is already optional as of Cargo 1.75, but this is currently equivalent to `version = "0.0.0"`.)
```

Comment:
As this is a "might", this should be covered in Unresolved Questions

Reply:
Moved to Unresolved Questions.